chaindataloader/binary.go

```diff
@@ -0,0 +1,227 @@
+
+	// Read data type.
+	dataTypeBytes := [8]byte{}
+	typeOfData, err := tlv.ReadVarInt(reader, &dataTypeBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining data type "+
+			"of file %w", err)
+	}
+
+	if dataType(typeOfData) != typeData {
+		return nil, fmt.Errorf("data type mismatch: got %v but expected %v",
+			dataType(typeOfData), typeData)
+	}
+
+	var dataTypeSize uint32
+	switch {
+	case typeData == blockHeaders:
+		dataTypeSize = headerfs.BlockHeaderSize
+	default:
+		return nil, fmt.Errorf("got unsupported dataType: %v", typeData)
+	}
+
+	// Read start height of block header file.
+	startBytes := [8]byte{}
+	start, err := tlv.ReadVarInt(reader, &startBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining start height "+
+			"of file %w", err)
+	}
+
+	// Read end height of block header file.
+	endBytes := [8]byte{}
+	end, err := tlv.ReadVarInt(reader, &endBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining end height "+
+			"of file %w", err)
+	}
+
+	// Read the bitcoin network, the headers in the header file belong to.
+	netBytes := [8]byte{}
+	chainChar, err := tlv.ReadVarInt(reader, &netBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining chain "+
+			"of file %w", err)
+	}
+
+	var chain wire.BitcoinNet
+	switch {
+	case chainChar == 1:
+		chain = wire.TestNet3
+	case chainChar == 2:
+		chain = wire.MainNet
+	case chainChar == 3:
+		chain = wire.SimNet
+	case chainChar == 4:
+		chain = wire.TestNet
+	default:
+		return nil, fmt.Errorf("read unsupported character (%d) for "+
+			"network of side-load source file", chainChar)
+	}
+
+	// obtain space occupied by metadata as initial offset
+	initialOffset, err := reader.Seek(0, io.SeekCurrent)
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to determine initial offset: "+
+			"%v", err)
+	}
+
+	return &binReader{
+		reader:        reader,
+		startHeight:   uint32(start),
+		endHeight:     uint32(end),
+		chain:         chain,
+		initialOffset: uint32(initialOffset),
+		dataTypeSize:  dataTypeSize,
+	}, nil
+}
+
+// NextBlockHeaders fetches headers from the reader. An empty slice is returned
+// if there are no headers left to fetch.
+func (b *binReader) NextBlockHeaders(n uint32) ([]*wire.BlockHeader, error) {
+	headers := make([]*wire.BlockHeader, n)
+
+	createData := func(headers []*wire.BlockHeader) func(data []byte,
+		i uint32) error {
+
+		return func(data []byte, i uint32) error {
+			var blockHeader wire.BlockHeader
+
+			headerReader := bytes.NewReader(data)
+
+			// Finally, decode the raw bytes into a proper bitcoin header.
+			if err := blockHeader.Deserialize(headerReader); err != nil {
+				return fmt.Errorf("error deserializing block header: %w",
+					err)
+			}
+
+			headers[i] = &blockHeader
+			return nil
+		}
+	}
+
+	err := b.readData(n, createData(headers))
+
+	if err != nil {
+		return nil, err
+	}
+	return headers, nil
+}
+
+func (b *binReader) readData(n uint32, createData func(data []byte,
```

Comment:
Missing a `godoc` comment.

Reply:
If you want to allow this to work for both the block header and the filter header, then I think we can decouple it from the struct (not a member), then use a type param, eg: 
```go

type neutrinoHeader interface {
     wire.BlockHeader | wire.FilterHeader
}
type headerDecoder[T neutrinoHeader] func(io.Reader) (T, error)

func readHeaders[T neutrinoHeader](numHeaders uint32, r io.Reader, decoder headerDecoder[T]) (T, error) {
    ....
}
```